torch/_cuda_sanitizer/_sanitizer.py

```diff
@@ -0,0 +1,490 @@
+
+
+class AccessType(enum.Enum):
+    READ = enum.auto()
+    WRITE = enum.auto()
+
+    def __str__(self):
+        return "reading from" if self is AccessType.READ else "writing to"
+
+
+@dataclass
+class Access:
+    type: AccessType
+    seq_num: SeqNum
+    stream: StreamId
+    operator: str
+    names: List[str]
+    stack_trace: List[str]
+
+
+class SynchronizationError(Exception):
+    """Base class for errors detected by CSAN."""
+
+    pass
+
+
+class UnsynchronizedAccessError(SynchronizationError):
+    def __init__(
+        self,
+        data_ptr: DataPtr,
+        allocation_trace: Optional[List[str]],
+        current_access: Access,
+        previous_access: Access,
+    ):
+        with io.StringIO() as message:
+            message.write(
+                textwrap.dedent(
+                    f"""\
+                    ============================
+                    CSAN detected a possible data race on tensor with data pointer {data_ptr}
+                    Access by stream {current_access.stream} during kernel:
+                    {current_access.operator}
+                    {current_access.type} argument: {', '.join(current_access.names)}
+                    With stack trace:
+                    """
+                )
+            )
+            message.write(f"{''.join(current_access.stack_trace)}\n")
+            message.write(
+                textwrap.dedent(
+                    f"""\
+                    Previous access by stream {previous_access.stream} during kernel:
+                    {previous_access.operator}
+                    {previous_access.type} argument: {', '.join(previous_access.names)}
+                    With stack trace:
+                    """
+                )
+            )
+            message.write(f"{''.join(previous_access.stack_trace)}\n")
+            if allocation_trace:
+                message.write(
+                    "Tensor was allocated with stack trace:\n"
+                    f"{''.join(allocation_trace)}"
+                )
+            else:
+                message.write("Trace for tensor allocation not found.")
+            super().__init__(message.getvalue())
+
+
+def format_log_message(message: str) -> str:
+    return " ".join(line.strip() for line in message.strip().splitlines())
+
+
+@dataclass
+class TensorInfo:
+    allocation_trace: Optional[List[str]]
+    reads: List[Access] = field(default_factory=list)
+    write: Optional[Access] = None
+
+
+class _TensorsAccessed:
+    def __init__(self):
+        self.accesses: Dict[DataPtr, TensorInfo] = {}
+
+    def ensure_tensor_exists(self, data_ptr: DataPtr) -> None:
+        if data_ptr not in self.accesses:
+            logger.info(
+                format_log_message(
+                    f"""
+                    Found tensor with pointer: {data_ptr}, but no matching tensor
+                    allocation in the trace. Backfilling the trace now.
+                    No action is necessary.
```

Comment:
So... why do we need this log message then?

Reply:
We've decided to log such occurrences, as something like this (missing the allocation event) should never occur during normal execution and could be a result of some obscure error.

Reply:
Then that's what the log message should say :)

Reply:
Now that CSAN is enabled by a function, I've added a warning that perhaps it is being enabled too late in the program, as that is now the most probable reason.